### PR TITLE
8305113: (tz) Update Timezone Data to 2023c

### PR DESCRIPTION
**Pull request by satoyoshiki**
Pleases review this PR. 
The PR includes the following changes.
- tzdata 2023c
- Hack code to deal with Cairo's DST end, which is same as done in 2014g([JDK-8049343](https://bugs.openjdk.org/browse/JDK-8049343)).  To work around the the limitation of JSR 310 compiler, where the time 24:00 is recognized as 0:00 in the previous day.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8305113](https://bugs.openjdk.org/browse/JDK-8305113): (tz) Update Timezone Data to 2023c

### Reviewers
 * [Naoto Sato](https://openjdk.org/census#naoto) (@naotoj - **Reviewer**)
 * [Andrew John Hughes](https://openjdk.org/census#andrew) (@gnu-andrew - **Reviewer**)
 * [Sergey Bylokhov](https://openjdk.org/census#serb) (@mrserb - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/13255/head:pull/13255` \
`$ git checkout pull/13255`

Update a local copy of the PR: \
`$ git checkout pull/13255` \
`$ git pull https://git.openjdk.org/jdk.git pull/13255/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 13255`

View PR using the GUI difftool: \
`$ git pr show -t 13255`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/13255.diff">https://git.openjdk.org/jdk/pull/13255.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/13255#issuecomment-1491121392)

### CODE REVIEW

**Comment by satoyoshiki** [comment]
Please review this PR. @naotoj

**Comment by naotoj** [comment]
IIUC, this PR is for `2023a` so this should be withdrawn. Instead, you may want to open a PR for https://bugs.openjdk.org/browse/JDK-8305113 which is for `2023c`?

**Comment by satoyoshiki** [comment]
> @satoyoshiki The issue `8304763` was not found in the list of additional solved issues. The list currently contains these issues: `8305113`

@naotoj Thanks for the catch.  I mistakenly put the wrong JBS id into `/issue` command.  `8304763` is now removed then the bot has got the review link removed from `JDK-8304763`.  Is this ok for you?

File: src/java.base/share/data/tzdata/northamerica [hunk 1]
```diff
@@ -1,4 +1,4 @@
-#
+src/java.base/share/data/tzdata/northamerica #
```
Comment (naotoj) [inline src/java.base/share/data/tzdata/northamerica:1]: ??? copy/paste error?

**Review by naotoj — CHANGES_REQUESTED** [review]
Looks good overall. Please fix a couple of editorial errors.

File: src/java.base/share/data/tzdata/antarctica [hunk 2]
```diff
@@ -1,4 +1,4 @@
-#
+ #
```
Comment (naotoj) [inline src/java.base/share/data/tzdata/antarctica:1]: Extra space

**Review by naotoj — CHANGES_REQUESTED** [review]

File: src/java.base/share/classes/sun/util/resources/TimeZoneNames.java [hunk 3]
```diff
@@ -1,5 +1,5 @@
 /*
- * Copyright (c) 1996, 2022, Oracle and/or its affiliates. All rights reserved.
+ * Copyright (c) 1996, 2033, Oracle and/or its affiliates. All rights reserved.
```
Comment (mrserb) [inline src/java.base/share/classes/sun/util/resources/TimeZoneNames.java:2]: Wrong year.

File: src/java.base/share/classes/sun/util/calendar/ZoneInfoFile.java [hunk 4]
```diff
@@ -726,8 +737,9 @@ private static ZoneInfo getZoneInfo(String zoneId,
                 checksum = (int)sum.getValue();
             }
         }
+
         return new ZoneInfo(zoneId, rawOffset, dstSavings, checksum, transitions,
-                            offsets, params, willGMTOffsetChange);
+                offsets, params, willGMTOffsetChange);
```
Comment (mrserb) [inline src/java.base/share/classes/sun/util/calendar/ZoneInfoFile.java:742]: do we need to change it here?

File: src/java.base/share/data/tzdata/northamerica (hunk 1 above)
Comment (satoyoshiki) [reply src/java.base/share/data/tzdata/northamerica:1]: It's likely. Thanks for the catch.

File: src/java.base/share/data/tzdata/antarctica (hunk 2 above)
Comment (satoyoshiki) [reply src/java.base/share/data/tzdata/antarctica:1]: Thanks!

File: src/java.base/share/classes/sun/util/resources/TimeZoneNames.java (hunk 3 above)
Comment (satoyoshiki) [reply src/java.base/share/classes/sun/util/resources/TimeZoneNames.java:2]: Thanks!

File: src/java.base/share/classes/sun/util/calendar/ZoneInfoFile.java (hunk 4 above)
Comment (satoyoshiki) [reply src/java.base/share/classes/sun/util/calendar/ZoneInfoFile.java:742]: No need.  It was reformatted by IntelliJ editor.  The original code is better from perspective of the OpenJDK's standard code style.

**Review by gnu-andrew — APPROVED** [review]
Looks good to me. Interesting that `TimeZoneNames.java` already has the `MSK` change for the other timezone that was changed, `Europe/Volgograd`

**Review by naotoj — APPROVED** [review]
LGTM

**Review by mrserb — APPROVED** [review]

**Pull request closed without merging**